google-cloud-spanner-cassandra/src/main/java/com/google/cloud/spanner/adapter/Launcher.java

```diff
@@ -66,120 +109,380 @@ public class Launcher {
+        if (userConfigs.getGlobalClientConfigs().getHealthCheckEndpoint() != null) {
+          healthCheckConfig =
+              HealthCheckConfig.fromEndpointString(
+                  userConfigs.getGlobalClientConfigs().getHealthCheckEndpoint());
+        }
+      } else {
+        globalSpannerEndpoint = ConfigConstants.DEFAULT_SPANNER_ENDPOINT;
+        globalEnableBuiltInMetrics = false;
+      }
+
+      List<ListenerConfig> listenerConfigs = new ArrayList<>();
+      for (ListenerConfigs listener : userConfigs.getListeners()) {
+        validateListenerConfig(listener);
+        listenerConfigs.add(
+            ListenerConfig.fromListenerConfigs(
+                listener, globalSpannerEndpoint, globalEnableBuiltInMetrics));
+      }
+
+      return new LauncherConfig(listenerConfigs, healthCheckConfig);
+    }
+
+    static LauncherConfig fromProperties(Map<String, String> properties)
+        throws UnknownHostException {
+      String databaseUri = properties.get(ConfigConstants.DATABASE_URI_PROP_KEY);
+      if (databaseUri == null) {
+        throw new IllegalArgumentException(
+            "Spanner database URI not set. Please set it using the '"
+                + ConfigConstants.DATABASE_URI_PROP_KEY
+                + "' property.");
+      }
+
+      ListenerConfig listenerConfig = ListenerConfig.fromProperties(properties);
+      HealthCheckConfig healthCheckConfig = HealthCheckConfig.fromProperties(properties);
+
+      return new LauncherConfig(Collections.singletonList(listenerConfig), healthCheckConfig);
+    }
+
+    private static void validateListenerConfig(ListenerConfigs listener) {
+      if (listener.getSpanner() == null
+          || Strings.isNullOrEmpty(listener.getSpanner().getDatabaseUri())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Listener '%s' on port %s must have a non-empty 'spanner.databaseUri' defined.",
+                listener.getName(), listener.getPort()));
+      }
+    }
+  }
+
+  /** Encapsulates the configuration for a single Adapter listener. */
+  private static final class ListenerConfig {
+    private final String databaseUri;
+    private final InetAddress hostAddress;
+    private final int port;
+    private final String spannerEndpoint;
+    private final int numGrpcChannels;
+    @Nullable private final Integer maxCommitDelayMillis;
+    private final boolean enableBuiltInMetrics;
+
+    private ListenerConfig(
+        String databaseUri,
+        InetAddress hostAddress,
+        int port,
+        String spannerEndpoint,
+        int numGrpcChannels,
+        @Nullable Integer maxCommitDelayMillis,
+        boolean enableBuiltInMetrics) {
+      this.databaseUri = databaseUri;
+      this.hostAddress = hostAddress;
+      this.port = port;
+      this.spannerEndpoint = spannerEndpoint;
+      this.numGrpcChannels = numGrpcChannels;
+      this.maxCommitDelayMillis = maxCommitDelayMillis;
```

Comment:
I would use a builder pattern to make the instantiation of configuration objects more readable, less error-prone and easy for future extension